runtime/Java/src/org/antlr/v4/runtime/CodePointCharStream.java

```diff
@@ -0,0 +1,297 @@
+/*
+ * Copyright (c) 2012-2017 The ANTLR Project. All rights reserved.
+ * Use of this file is governed by the BSD 3-clause license that
+ * can be found in the LICENSE.txt file in the project root.
+ */
+
+package org.antlr.v4.runtime;
+
+import org.antlr.v4.runtime.misc.Interval;
+
+import java.nio.charset.StandardCharsets;
+
+/**
+ * Alternative to {@link ANTLRInputStream} which treats the input
+ * as a series of Unicode code points, instead of a series of UTF-16
+ * code units.
+ *
+ * Use this if you need to parse input which potentially contains
+ * Unicode values > U+FFFF.
+ */
+public abstract class CodePointCharStream implements CharStream {
+	protected final int size;
+	protected final String name;
+
+	// To avoid lots of virtual method calls, we directly access
+	// the state of the underlying code points in the
+	// CodePointBuffer.
+	protected int position;
+
+	// Use the factory method {@link #fromBuffer(CodePointBuffer)} to
+	// construct instances of this type.
+	private CodePointCharStream(int position, int remaining, String name) {
+		// TODO
+		assert position == 0;
+		this.size = remaining;
+		this.name = name;
+		this.position = 0;
+	}
+
+	// Visible for testing.
+	abstract Object getInternalStorage();
+
+	/**
+	 * Constructs a {@link CodePointCharStream} which provides access
+	 * to the Unicode code points stored in {@code codePointBuffer}.
+	 */
+//	public static CodePointCharStream fromBuffer(CodePointBuffer codePointBuffer) {
+//		return fromBuffer(codePointBuffer, UNKNOWN_SOURCE_NAME);
+//	}
+
+	/**
+	 * Constructs a named {@link CodePointCharStream} which provides access
+	 * to the Unicode code points stored in {@code codePointBuffer}.
+	 */
+	public static CodePointCharStream fromBuffer(CodePointBuffer codePointBuffer, String name) {
+		// Java lacks generics on primitive types.
+		//
+		// To avoid lots of calls to virtual methods in the
+		// very hot codepath of LA() below, we construct one
+		// of three concrete subclasses.
+		//
+		// The concrete subclasses directly access the code
+		// points stored in the underlying array (byte[],
+		// char[], or int[]), so we can avoid lots of virtual
+		// method calls to ByteBuffer.get(offset).
+		switch (codePointBuffer.getType()) {
+			case BYTE:
+				return new CodePoint8BitCharStream(
+						codePointBuffer.position(),
+						codePointBuffer.remaining(),
+						name,
+						codePointBuffer.byteArray(),
+						codePointBuffer.arrayOffset());
+			case CHAR:
+				return new CodePoint16BitCharStream(
+						codePointBuffer.position(),
+						codePointBuffer.remaining(),
+						name,
+						codePointBuffer.charArray(),
+						codePointBuffer.arrayOffset());
+			case INT:
+				return new CodePoint32BitCharStream(
+						codePointBuffer.position(),
+						codePointBuffer.remaining(),
+						name,
+						codePointBuffer.intArray(),
+						codePointBuffer.arrayOffset());
+		}
+		throw new UnsupportedOperationException("Not reached");
```

Comment:
I guess the problem is how to recreate the 3 CharStreams here in a way that works in GWT.